src/PartitionLoops.cpp

```diff
@@ -17,748 +19,652 @@ using std::string;
-            debug(3) << "Attempted to make this condition false: " << condition << " Got: " << new_condition << "\n";
-            if (is_zero(new_condition)) {
-                if (!found_simplification_in_children && tight) {
-                    Replacement r = {op->condition, const_true()};
-                    if (min_vals.size() > old_num_min_vals) {
-                        prologue_replacements.push_back(r);
-                    }
-                    if (max_vals.size() > old_num_max_vals) {
-                        epilogue_replacements.push_back(r);
-                    }
-                }
-                return false_value;
-            } else {
-                return SelectOrIf::make(new_condition, true_value, false_value);
+            if (interval_has_upper_bound(s.interval) &&
+                upper_bound_is_tight) {
+                internal_assert(s.tight);
+                Simplification s2 = s;
+                s2.condition = !s2.condition;
+                std::swap(s2.likely_value, s2.unlikely_value);
+                epilogue_simps.push_back(s2);
             }
-        } else if (condition.same_as(op->condition) &&
-                   true_value.same_as(orig_true_value) &&
-                   false_value.same_as(orig_false_value)) {
-            return op;
-        } else {
-            return SelectOrIf::make(condition, true_value, false_value);
         }
-    }
 
-    void visit(const Select *op) {
-        expr = visit_select_or_if(op, op->true_value, op->false_value);
-    }
+        // Simplify each section of the loop.
+        Stmt simpler_body = MakeSimplifications(middle_simps).mutate(body);
+        Stmt prologue = MakeSimplifications(prologue_simps).mutate(body);
+        Stmt epilogue = MakeSimplifications(epilogue_simps).mutate(body);
 
-    void visit(const IfThenElse *op) {
-        stmt = visit_select_or_if(op, op->then_case, op->else_case);
-    }
+        bool make_prologue = !equal(prologue, simpler_body);
+        bool make_epilogue = !equal(epilogue, simpler_body);
 
-    void visit(const Let *op) {
-        Expr value = mutate(op->value);
-        Expr body;
+        // Recurse on the middle section.
+        simpler_body = mutate(simpler_body);
 
-        bound_vars.push(op->name, value);
-        body = mutate(op->body);
-        bound_vars.pop(op->name);
+        // Construct variables for the bounds of the simplified middle section
+        Expr min_steady = op->min, max_steady = op->extent + op->min;
+        Expr prologue_val, epilogue_val;
+        string prologue_name = unique_name(op->name + ".prologue", false);
+        string epilogue_name = unique_name(op->name + ".epilogue", false);
 
-        if (value.same_as(op->value) && body.same_as(op->body)) {
-            expr = op;
-        } else {
-            expr = Let::make(op->name, value, body);
-        }
-    }
+        if (make_prologue) {
+            // They'll simplify better if you put them in lexicographic order
```

Comment:
Can you say why this is? It's an interesting comment. Do the simplifications conflict with each other?